appinventor/components/src/com/google/appinventor/components/runtime/Serial.java

```diff
@@ -0,0 +1,109 @@
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.components.runtime;
+
+import android.content.Context;
+import android.util.Log;
+
+import com.physicaloid.lib.Physicaloid;
+
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.annotations.DesignerComponent;
+import com.google.appinventor.components.annotations.SimpleEvent;
+import com.google.appinventor.components.annotations.SimpleFunction;
+import com.google.appinventor.components.annotations.SimpleObject;
+import com.google.appinventor.components.annotations.UsesLibraries;
+import com.google.appinventor.components.common.YaVersion;
+
+import java.io.UnsupportedEncodingException;
+
+@DesignerComponent(version = YaVersion.SERIAL_COMPONENT_VERSION,
+    description = "Serial component which can be used to connect to devices like Arduino",
+    category = ComponentCategory.CONNECTIVITY,
+    nonVisible = true,
+    iconName = "images/arduino.png")
+
+@SimpleObject
+@UsesLibraries(libraries = "physicaloid.jar")
+public class Serial extends AndroidNonvisibleComponent implements Component {
+  private static final String LOG_TAG = "Serial Component";
+
+  private Context context;
+
+  Physicaloid mPhysicaloid;
+
+  private int baudRate = 9600;
+
+  public Serial(ComponentContainer container) {
+    super(container.$form());
+    context = container.$context();
+    Log.d(LOG_TAG, "Created");
+  }
+
+  @SimpleFunction(description = "Initializes serial connection.")
+  public void InitializeSerial() {
+    mPhysicaloid = new Physicaloid(context);
+    Log.d(LOG_TAG, "Initialized");
+  }
+
+  @SimpleFunction(description = "Opens serial connection. Returns true when opened.")
+  public boolean OpenSerial() {
+    Log.d(LOG_TAG, "Opening connection");
+    return mPhysicaloid.open();
+  }
+
+  @SimpleFunction(description = "Closes serial connection. Returns true when closed.")
+  public boolean CloseSerial() {
+    Log.d(LOG_TAG, "Closing connection");
+    return mPhysicaloid.close();
+  }
+
+  @SimpleFunction(description = "Sets a new baud rate. Default is 9600 bps.")
+  public void BaudRate(int baudRate) {
+    this.baudRate = baudRate;
+    mPhysicaloid.setBaudrate(baudRate);
+    Log.d(LOG_TAG, "Baud Rate: " + baudRate);
+  }
+
+  @SimpleFunction(description = "Reads data from serial.")
+  public void ReadSerial() {
+    byte[] buf = new byte[256];
+    boolean success = true;
+    String data = "";
+
+    if (mPhysicaloid.read(buf) > 0) {
+      try {
+        data = new String(buf, "UTF-8");
+      } catch (UnsupportedEncodingException mEr) {
+        success = false;
+        Log.e(LOG_TAG, mEr.getMessage());
+      }
+    } else {
+      success = false;
+    }
+
+    AfterReadSerial(success, data);
+  }
+
+  @SimpleFunction(description = "Writes given data to serial.")
+  public void WriteSerial(String writeDataSerial) {
```

Comment:
I wonder if it might also make sense to have a "PrintSerial" function that automatically includes the \n.

Reply:
Added new method. However, I think that, if the user wants to add "\n" he will add it in blocks. We may get questions about why having both blocks.